systemverilog-plugin/UhdmAst.cc

```diff
@@ -357,51 +357,58 @@ static void add_multirange_wire(AST::AstNode *node, std::vector<AST::AstNode *>
     }
 }
 
-static size_t add_multirange_attribute(AST::AstNode *wire_node, const std::vector<AST::AstNode *> ranges)
+// Sets the `wire_node->multirange_dimensions` attribute and returns the total sizes of packed and unpacked ranges.
+static std::pair<size_t, size_t> set_multirange_dimensions(AST::AstNode *wire_node, const std::vector<AST::AstNode *> packed_ranges,
+                                                           const std::vector<AST::AstNode *> unpacked_ranges)
 {
     // node->multirange_dimensions stores dimensions' offsets and widths.
     // It shall have even number of elements.
     // For a range of [A:B] it should be appended with {min(A,B)} and {max(A,B)-min(A,B)+1}
     // For a range of [A] it should be appended with {0} and {A}
 
-    size_t size = 1;
-    for (size_t i = 0; i < ranges.size(); i++) {
-        log_assert(AST_INTERNAL::current_ast_mod);
-        if (ranges[i]->children.size() == 1) {
-            ranges[i]->children.push_back(ranges[i]->children[0]->clone());
-        }
-        simplify_sv(ranges[i], wire_node);
-        while (simplify(ranges[i], true, false, false, 1, -1, false, false)) {
-        }
-        // this workaround case, where yosys doesn't follow id2ast and simplifies it to resolve constant
-        if (ranges[i]->children[0]->id2ast) {
-            simplify_sv(ranges[i]->children[0]->id2ast, ranges[i]->children[0]);
-            while (simplify(ranges[i]->children[0]->id2ast, true, false, false, 1, -1, false, false)) {
+    auto calc_range_size = [&wire_node](const std::vector<AST::AstNode *> ranges) -> size_t {
```

Comment:
I think it is better to copy ``wire_node`` pointer (I'm not sure about this, but as it is not const, this makes sure that we won't change it inside lambda, that will be visible outside of lambda) and it is better to take ``std::vector<AST::AstNode *>`` as a reference.
```suggestion
    auto calc_range_size = [wire_node](const std::vector<AST::AstNode *>& ranges) -> size_t {
```

Reply:
Alright, I've applied this.